# 4. エラー処理

Xオブジェクトのメソッドがエラーを生成した場合、メソッドの種類に応じて2つの異なる方法のいずれかで処理されます。エラーはXlibのXエラーオブジェクトで表現されます。

## データを返さないメソッドの場合

メソッドが **データを返さない** 場合、エラーはメソッドが戻った後に検出される可能性があります。データを返さないすべてのメソッドには `onerror` というパラメータがあり、このメソッドによって生成されるエラーのエラーハンドラーを提供するために使用できます。

### エラーハンドラーの動作

エラーハンドラーは2つの引数で呼び出されます：
- エラーオブジェクト
- 低レベルリクエストオブジェクト

**重要**: エラーハンドラーは、Xオブジェクトメソッドを呼び出してはいけません。それが必要な場合、エラーハンドラーは後で取得するためにエラーを保存しておく必要があります。この目的のために `error.CatchError` クラスが提供されています。

### デフォルトエラーハンドラー

エラーを生成するメソッドを呼び出す際にエラーハンドラーが指定されていない場合、エラーはデフォルトエラーハンドラーに渡されます。デフォルトエラーハンドラーが指定されていない場合、エラーは単に `sys.stderr` に出力されます。

## データを返すメソッドの場合

メソッドが **データを返す** 場合、エラーによって有効なデータを返すことができなくなります。この場合、エラーオブジェクトを例外値として例外が発生します。エラーハンドラーには渡されません。

## 関連項目

- [Xエラークラス](04-1_Xエラークラス.md)
- [CatchError](04-2_CatchError.md)